src/const-eval.md

```diff
@@ -20,17 +20,50 @@ Additionally constant evaluation can be used to reduce the workload or binary
 size at runtime by precomputing complex operations at compiletime and only
 storing the result.
 
+All uses of constant evaluation can either be categorized as "influencing the type system"
+(array lengths, enum variant discriminants, const generic parameters), or as solely being
+done to precompute expressions to be used at runtime.
+
 Constant evaluation can be done by calling the `const_eval_*` functions of `TyCtxt`.
 They're the wrappers of the `const_eval` query.
 
+`static` initializers must use the `eval_static_initializer` function. All other functions
```

Comment:
This comes out of the blue... and it is strange that we are naming this function out of all of them. Given teh big paragraph about valtrees, I'd have expected something like(names are made up as I am not always sure what the real thing is called)

> Constant evaluation can be done by calling the `const_eval_*` functions of `TyCtxt`. They're the wrappers of the `const_eval` query.
> * `const_eval_valtree` evaluates a constant to a valtree, so the result value can be further inspected by the compiler.
> * `const_eval_const` evaluate a constant to an "opaque blob" containing its final value; this is only useful for codegen backends and the CTFE evaluator engine itself.
> `eval_static_initializer` specifically computes the initial values of a static. Statics are special; all other functions do not represent statics correctly and have thus assertions preventing their use on statics.

However, the static special case should also disappear once valtrees are properly done, right? All of this "normalization" magic should not be needed any more.